From: Michael Vogt <mvo@ubuntu.com>
Date: Fri, 20 Dec 2019 16:45:51 +0100
Subject: [PATCH] add support for --extrausers for chfn

This add support for --extrausers to the chfn tool.
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1495580

[Changelog]
- This patch has been refreshed to fix warning 'implicit function declaration'
  Hector Cao <hector.cao@canonical.com>
---
 src/chfn.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/src/chfn.c b/src/chfn.c
index 834a0a2..67ce7bc 100644
--- a/src/chfn.c
+++ b/src/chfn.c
@@ -38,6 +38,12 @@
 #include "string/strdup/xstrdup.h"
 #include "string/strtok/stpsep.h"
 
+#include "shadowio.h"
+#include "groupio.h"
+#ifdef SHADOWGRP
+#include "sgroupio.h"
+#endif
+#include "subordinateio.h"
 
 /*
  * Global variables.
@@ -57,6 +63,11 @@ static bool hflg = false;		/* -h - set home phone number        */
 static bool oflg = false;		/* -o - set other information        */
 static bool pw_locked = false;
 
+#ifndef EXTRAUSERS_OPT
+#define EXTRAUSERS_OPT 100000
+#endif
+static bool use_extrausers = false;
+
 /*
  * External identifiers
  */
@@ -111,6 +122,7 @@ usage (int status)
 	(void) fputs (_("  -R, --root CHROOT_DIR         directory to chroot into\n"), usageout);
 	(void) fputs (_("  -u, --help                    display this help message and exit\n"), usageout);
 	(void) fputs (_("  -w, --work-phone WORK_PHONE   change user's office phone number\n"), usageout);
+	(void) fputs (_("      --extrausers              Use the extra users database\n"), usageout);
 	(void) fputs ("\n", usageout);
 	exit (status);
 }
@@ -260,6 +272,7 @@ static void process_flags (int argc, char **argv)
 		{"root",       required_argument, NULL, 'R'},
 		{"help",       no_argument,       NULL, 'u'},
 		{"work-phone", required_argument, NULL, 'w'},
+		{"extrausers", no_argument, NULL, EXTRAUSERS_OPT},
 		{NULL, 0, NULL, '\0'}
 	};
 
@@ -273,6 +286,9 @@ static void process_flags (int argc, char **argv)
 	while ((c = getopt_long (argc, argv, "f:h:o:r:R:uw:",
 	                         long_options, NULL)) != -1) {
 		switch (c) {
+		case EXTRAUSERS_OPT:
+			use_extrausers = true;
+			break;
 		case 'f':
 			if (!may_change_field ('f')) {
 				fprintf (stderr,
@@ -645,6 +661,18 @@ int main (int argc, char **argv)
 	/* parse the command line options */
 	process_flags (argc, argv);
 
+	if (use_extrausers) {
+		pw_setdbname(EXTRAUSERS_PASSWD_FILE);
+		spw_setdbname(EXTRAUSERS_SHADOW_FILE);
+		gr_setdbname(EXTRAUSERS_GROUP_FILE);
+		/* TODO expose this information in other tools */
+		sub_uid_setdbname(EXTRAUSERS_SUBUID_FILE);
+		sub_gid_setdbname(EXTRAUSERS_SUBGID_FILE);
+		#ifdef SHADOWGRP
+		sgr_setdbname(EXTRAUSERS_SHADOWGROUP_FILE);
+		#endif
+	}
+
 	/*
 	 * Get the name of the user to check. It is either the command line
 	 * name, or the name getlogin() returns.
